src/compiler/checker.ts

```diff
@@ -17698,41 +17989,43 @@ namespace ts {
-                    typeAsIterable.iterableElementType = (<GenericType>type).typeArguments[0];
-                }
-                else {
-                    const iteratorFunction = getTypeOfPropertyOfType(type, getPropertyNameForKnownSymbolName("iterator"));
-                    if (isTypeAny(iteratorFunction)) {
-                        return undefined;
-                    }
+            if (typeAsIterable.iteratedTypeOfIterable) {
+                return typeAsIterable.iteratedTypeOfIterable;
+            }
 
-                    const iteratorFunctionSignatures = iteratorFunction ? getSignaturesOfType(iteratorFunction, SignatureKind.Call) : emptyArray;
-                    if (iteratorFunctionSignatures.length === 0) {
-                        if (errorNode) {
-                            error(errorNode, Diagnostics.Type_must_have_a_Symbol_iterator_method_that_returns_an_iterator);
-                        }
-                        return undefined;
-                    }
+            // As an optimization, if the type is instantiated directly using the
+            // globalIterableType (Iterable<T>) or globalIterableIteratorType (IterableIterator<T>),
+            // then just grab its type argument.
+            if (isReferenceToType(type, getGlobalIterableType(/*reportErrors*/ false)) ||
+                isReferenceToType(type, getGlobalIterableIteratorType(/*reportErrors*/ false))) {
+                return typeAsIterable.iteratedTypeOfIterable = (<GenericType>type).typeArguments[0];
+            }
 
-                    typeAsIterable.iterableElementType = getElementTypeOfIterator(getUnionType(map(iteratorFunctionSignatures, getReturnTypeOfSignature), /*subtypeReduction*/ true), errorNode);
+            const propertyName = getPropertyNameForKnownSymbolName("iterator");
+            const iteratorMethod = getTypeOfPropertyOfType(type, propertyName);
+            if (isTypeAny(iteratorMethod)) {
+                return undefined;
+            }
+
+            const iteratorMethodSignatures = iteratorMethod ? getSignaturesOfType(iteratorMethod, SignatureKind.Call) : emptyArray;
+            if (iteratorMethodSignatures.length === 0) {
+                if (errorNode) {
+                    error(errorNode, Diagnostics.Type_must_have_a_Symbol_iterator_method_that_returns_an_iterator);
                 }
+                return undefined;
             }
 
-            return typeAsIterable.iterableElementType;
+            return typeAsIterable.iteratedTypeOfIterable = getIteratedTypeOfIterator(getUnionType(map(iteratorMethodSignatures, getReturnTypeOfSignature), /*subtypeReduction*/ true), errorNode);
         }
 
         /**
-         * This function has very similar logic as getElementTypeOfIterable, except that it operates on
+         * This function has very similar logic as getIteratedTypeOfIterable, except that it operates on

```

Comment:
The `getIteratedTypeOf*` functions aren't repetitive, just similarly structured. Each function has a different responsibility. Having a single `getIteratedType` function with various parameters to control its behavior would conflate all of those responsibilities and result in a very large function.

* `getIteratedTypeOfIterator` - gets the iterated type from an `Iterator<T>`-like type, with a fast path for type references to `Iterator<T>`.
* `getIteratedTypeOfIterable` - gets the iterated type from an `Iterable<T>`-like type, with a fast path for type references to `Iterable<T>` and `IterableIterator<T>`.  If the type has a `Symbol.iterator` method, it relies on `getIteratedTypeOfIterator`.
* `getIteratedTypeOfIterableIterator` - gets the iterated type from an `Iterable<T>`-like type or an `Iterator<T>`-like type. The order it resolves the type is important as this is used to verify the return type of a generator function.

* `getIteratdTypeOfAsyncIterator` - While similar to `getIteratedTypeOfIterator`, this function specifically targets `AsyncIterator<T>` and gets the promised type of the `next()` method. This function uses a different cache location for the resulting type, so any attempt to merge these functions would likely result in a significant amount of bifurcation.
* `getIteratedTypeOfAsyncIterable` - While similar to `getIteratedTypeOfIterable`, this function specifically targets `AsyncIterable<T>`. As with `getIteratedTypeOfAsyncIterator`, merging this function with its non-async counterpart would result in significant bifurcation due to how conflicts are cached.

* `getIteratedTypeOfIterableOrElementTypeOfArray` - used to get the iterated type in a for-of statement for `--target es5`.
* `getIteratedTypeOfIterableOrElementTypeOfArrayOrString` - used as part of checking the iterated type of the right-hand-side of a for-of statement where `string` may be allowed.
* `getIteratedTypeOrElementType` - conditionally calls `getIteratedTypeOfIterable`, `getIteratedTypeOfIterableOrElementTypeOfArray` or `getIteratedTypeOfIterableOrElementTypeOfArrayOrString` based on passed in requirements.

While this may seem like a "thicket", each function has a specific responsibility and I think is generally more maintainable than a monolithic function that tries to serve far too many needs.